### PR TITLE
mouse-actions-gui: init at 0.4.4

### LINKED ISSUE #269212: Package request: mouse-actions-gui
**Project description**

mouse-actions allows to execute some commands from mouse events such as:

- Clicks / wheel on the side / corners of the screen,
- Or drawing shapes.

**Obs:** 
- There is the no-gui package request here: https://github.com/NixOS/nixpkgs/issues/264904
- This request is for implementing the `mouse-actions-gui` element so we can configure `mouse-actions`.

**Metadata**

* homepage URL: https://github.com/jersou/mouse-actions
* source URL: https://github.com/jersou/mouse-actions
* license: mit, bsd, gpl2+ , ...
* platforms: unix, linux, darwin, ...

**Priorities**

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by TomaSajt**
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/269212
Connected to https://github.com/NixOS/nixpkgs/issues/264904

Mostly based on the GUI-less solution ([link](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/mo/mouse-actions/package.nix))

This PR adds 1 package: `mouse-actions-gui`
There is a GUI-less version called `mouse-actions`, but this package also has all of that functionality (can be run GUI-less).

The package is a `tauri` app:
I tried to use `buildRustPackage`, like I did for a previous `tauri` app, but for some reason `vite` wanted to connect to the development server even after building. (probably some environment variables I couldn't figure out)
Instead I used `cargo-tauri` to build it as an unwrapped `.deb` package. This actually also takes care of renaming the installed files and the installing of `.desktop` and icons files.

This was inspired by how the [`pot` package](https://github.com/NixOS/nixpkgs/blob/40de5938426b06886a2b441142031c29e0eb8dc1/pkgs/applications/misc/pot/default.nix#L121) did it

I added the necessary `udev` rules based off of the GUI-less solution (@rgri).
(You'll have to add the package to `services.udev.packages` in your configs for the rules to register)
Not exactly sure what the 2nd rule does, as it seems to work without it, but I'll also put it there for parity's sake.

I edited the `programs.mouse-actions` nixod module to be able to take in a different package.

I have also added a systemd service which can autorun the program on launch. It has to do some hacky solutions to use the user's PATH variable, but it works. It is locked behind the `.autorun` option.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by rgri** [comment]
Hi! Thanks for the help, I was planning on looking at this eventually but forgot lol.
1. The [second rule](https://github.com/NixOS/nixpkgs/pull/265056/commits/71717ff114eec1429bf0c7ad95b3175f3f28cf10#diff-5276b4aab67c7d0921968cda3c904b06300cdab0f5632fb57308762bc3e6f86bR38) (so others can follow) was included as part of the [upstream requirements](https://github.com/jersou/mouse-actions#requirement-), specifically "To use the main feature "grab event", you need to have the read&write permission on `/dev/input/event*`." Interesting that you can run it without that- maybe I didn't do thorough testing.
2. Instead of requiring users to manually add this package to `services.udev.packages` you should probably write a module [like mine](https://github.com/NixOS/nixpkgs/pull/265056/commits/71717ff114eec1429bf0c7ad95b3175f3f28cf10#diff-4feb38f0912e4e082f731b5a02366370598b93d44774688a7ad902e2c47445b4) which does it for them on `enable = true;`. This has precendent in the package [glasgow](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/hardware/glasgow.nix#L20).

**Comment by rgri** [comment]
Hi hi. #265056 just got merged. Would you be interested in working on this together? If you aren't, I'd be happy to take it from here.

**Comment by TomaSajt** [comment]
> Hi hi. #265056 just got merged. Would you be interested in working on this together? If you aren't, I'd be happy to take it from here.

I'd be up to discuss how to deal with the duality of the two packages.

One option would be to keep them separate, as `mouse-actions` and `mouse-actions-gui`.
But then most of basic package data would be the same, so that's a bit of code duplication.

Another option would be to have only `mouse-actions`, but you can enable/disable the gui by overriding the package, with something like this:
`mouse-actions.override { withGui = false/true; }`
This would need the two packages to be combined into one file. They are quite different, as I the gui package is using `buildNpmPackage` with `cargoSetupHook` and `cargo-tauri` while the non-gui package is using the normal `buildRustPackage`.

Whichever option is chosen, we'd need to perhaps be able to override the `mouse-actions` package in `programs.mouse-actions`
Currently it's hard coded to `pkgs.mouse-actions` but in most other program modules, theres a `.package` attribute, in this case `programs.mouse-actions.package` would be the one needed to be added.

There's also the question of which package should be the default for the service (or for the combined package)

**Comment by waldman** [comment]
IMHO, for the sake of simplicity I would vote for two packages... From my experience some code duplication is not an issue if it makes things simpler/easier. (But this is just my "two cents" on the matter). =)

**Comment by waldman** [comment]
Hey all! =)

Any news on this one?

**Comment by TomaSajt** [comment]
> Hey all! =)
> 
> Any news on this one?

Hey, thanks for the reminder. I've forgotten about this a little bit.
I'll try to get someone to look at this PR.

**Review by dblsaiko — CHANGES_REQUESTED** [review]
Seems to work, at least the GUI. I haven't tested the actual functionality.

As @rgri said, please add a NixOS module that adds this to system packages and udev packages so people can just write `programs.mouse-actions-gui.enable = true;`. Take a look at nixos/modules/programs/dmrconfig.nix for what I'm thinking of.

(Also personally, especially if there's multiple build systems involved instead of mkDerivation wrappers like buildNpmPackage I would just use mkDerivation and add the necessary cargo and npm hooks manually, that composes better. Then you can also use finalAttrs. Just a suggestion, keep it like this if you want to.)

Result of `nixpkgs-review pr 269518` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mouse-actions-gui</li>
  </ul>
</details>

**Comment by TomaSajt** [comment]
I have added a NixOS module in a new commit (untested), however I am not really happy with this current solution, as there are now 2 modules for the (almost) same program (`mouse-actions` and `mouse-actions-gui`).
It might be better to have a `programs.mouse-actions.package` attribute, which can be set to `mouse-actions-gui`. (Or maybe that should be the default, and you could supply use the non-gui version)

**Comment by 2xsaiko** [comment]
It should also add it to systemPackages, I think. I'll have to test whether it adds a desktop entry and the binary to PATH like this, but I don't think so.

As for the duplicate module, that's fair. How is this used along with the other tool usually? Do you use one or the other, or is it useful to have both at the same time?

**Comment by TomaSajt** [comment]
The gui app is a superset of the non-gui app.
I am working on updating the mouse-actions module to:
- add the package to systemPackages
- create a oneshot systemd user service to launch the program
- add pkgs.programs.mouse-actions.package for overriding.
  - maybe instead of this we could have a gui=true; option

**Comment by waldman** [comment]
@TomaSajt thanks man! Really really thanks! =))

**Comment by 2xsaiko** [comment]
> The gui app is a superset of the non-gui app.

Okay, yes, then I would have only the one already existing module and just swap out the package.

**Comment by TomaSajt** [comment]
Okay, so I've added a systemd user service to the module (I will put it behind a configuration flag later), which ALMOST works, however whenever I run it it turns of my wifi support (????, how?).
I'll check this out on my desktop PC later today.
If anyone knows what better `.target`s to use, or how to better setup the service dependency tree, please do tell me.

**Comment by 2xsaiko** [comment]
graphical-session.target is fine (though I think it should be bindsTo rather than requires). Are there any related messages in journalctl? Also, how are you setting up wifi?

**Comment by TomaSajt** [comment]
> > Anyways, I realize that because this is started as a systemd service, it has no idea about `$PATH` so it can't open apps without full paths to them, which is not the most convenient thing.
> 
> Try setting `environment.PATH = mkForce null;` in the service. It's NixOS that overrides it by default, user services usually inherit the environment.

I thought I found the right place where it's being handled
https://github.com/NixOS/nixpkgs/blob/ab14a429cb1a6cc7b7b843a8ad2661ad1570e57c/nixos/lib/systemd-lib.nix#L345-L359

But, it turns out this isn't really where it path comes from. It does, if you don't override anything, but if you do `path = lib.mkForce []`, then because of the `mkIf` above, `environment` won't get set, which you'd thing is a good thing. However there's an actual default probably compiled into `systemd`, which is the same as this list, but can't be overridden. The same thing happens if I use `mkForce null` on `environment`.

I am not sure what to do next...

**Comment by TomaSajt** [comment]
Wait, I might have actually came up with something working!!

**Comment by TomaSajt** [comment]
It works! Seems like `PassEnvironment` had to be used together with either `mkForce` attempt.

File: nixos/modules/programs/mouse-actions.nix [hunk 1]
```diff
@@ -3,13 +3,39 @@
...
+    enable = lib.mkEnableOption ''
+      mouse-actions udev rules. This is a prerequisite for using mouse-actions without being root.
+    '';
+    package = lib.mkOption {
+      type = lib.types.package;
+      example = lib.literalExpression "pkgs.mouse-actions-gui";
+      default = pkgs.mouse-actions;
+      defaultText = lib.literalExpression "pkgs.mouse-actions";
+      description = "Which mouse-actions package to use.";
+    };
```
Comment (2xsaiko) [inline nixos/modules/programs/mouse-actions.nix:17]:
```suggestion
    package = lib.mkPackageOption pkgs "mouse-actions" {
      example = "pkgs.mouse-actions-gui";
    };
```

File: nixos/modules/programs/mouse-actions.nix [hunk 2]
```diff
@@ -3,13 +3,39 @@
...
 in
-  {
-    options.programs.mouse-actions = {
-      enable = lib.mkEnableOption ''
-        mouse-actions udev rules. This is a prerequisite for using mouse-actions without being root.
+{
+  options.programs.mouse-actions = {
+    enable = lib.mkEnableOption ''
+      mouse-actions udev rules. This is a prerequisite for using mouse-actions without being root.
+    '';
```
Comment (2xsaiko) [inline nixos/modules/programs/mouse-actions.nix:10]: This should not say that it specifically adds the udev rules anymore. Maybe simply

```suggestion
    enable = lib.mkEnableOption "mouse-actions";
```

**Review by dblsaiko — CHANGES_REQUESTED** [review]
Nice! I didn't know PassEnvironment was a thing.

Couple of things, then it should be good.

File: nixos/modules/programs/mouse-actions.nix (hunk 1 above)
Comment (TomaSajt) [reply nixos/modules/programs/mouse-actions.nix:17]: Great, didn't know about this. I don't see other packages using the `pkgs.` prefix when using this so I'll omit it too.

**Comment by TomaSajt** [comment]
Maybe there should be some message instructing the user to add themselves to the `uinput` group to use this.

**Review by dblsaiko — APPROVED** [review]

**Comment by TomaSajt** [comment]
Just realized that I messed up the second line of the udev rule

**Comment by TomaSajt** [comment]
> @TomaSajt can you please point me to some NixOS documentation on how to use this locally until it is available in unstable?

Well, I asked around the Matrix chat a week ago, but nobody mentioned a really easy way.

I usually just create a `mouse-actions-module.nix` file in my configuration directory, copy over the PR contents, and rename the module from `programs.mouse-actions` to `programs.mouse-actions-dev` (all 2 places) to avoid naming conflicts. Then include the file in my configuration using `imports = [ ./mouse-actions-module.nix ];`.

Now you can do
`programs.mouse-actions-dev.enable = true;` to install the program
and also do
`programs.mouse-actions-dev.autorun = true;` to enable it to launch at startup. You'll need to restart your PC/session beforehand, probably.

If you also want to use `mouse-actions-gui`, you can either create some kind of overlay (any way you want to or are able to), and set
`programs.mouse-actions-dev.package = pkgs.mouse-actions-gui;`

Or you could copy the derivation into a new `mouse-actions-gui.nix` file in your configuration (you'll need to copy the udev rules file too) and do `programs.mouse-actions-dev.package = pkgs.callPackage ./mouse-actions-gui.nix { };`

**Comment by TomaSajt** [comment]
> Thank you for the detailed explanation!
> 
> I will follow what you suggested and see how far I can go. So far it seems I need to copy `nixos/modules/programs/mouse-actions.nix` file locally and enable/use that. I did not get the last few points about gui option and udev rules and I'm not sure how to transfer/use files in `pkgs/by-name/mo/mouse-actions-gui/*` locally. But I will re-read, explore, and try when I get home. Thank you @TomaSajt

I went ahead and did what I wrote down, here it is in my config: https://github.com/TomaSajt/nixfiles/commit/d4e35046cd0b568b15fca0a141d3693a29cbc3c2

**Comment by TomaSajt** [comment]
> This is not relevant but in case you've also run into it, how do you setup a `mouse-button-hold` + `scroll-wheel-up/down` trigger? I can't seem to get even the normal `mouse-wheel-up` working, and those window zones are not working either. I tried using app-image on ubuntu but no luck there either. Maybe mouse-actions does not support all the things `easystroke` did.

Sorry, I don't have much experience with this program either. If this issue is not nixpkgs-specific, then perhaps you can get more support from the upstream repo.

**Comment by TomaSajt** [comment]
> Hi all! o/
> 
> What still need to be done on this PR? Maybe I could help?

Hey! I haven't really received much feedback about the NixOS module yet, so if you could try if the `autorun = true` option works for you, I'd appreciate it.

**Comment by TomaSajt** [comment]
I've come up with a small nixos module which you can import into your on config that applies the changes to the module and enables autorun and mouse-actions-gui

```nix
{ ... }:

let
  nixpkgs-with-magui = fetchTarball {
    url = "https://github.com/TomaSajt/nixpkgs/archive/mouse-actions-gui.tar.gz";
    sha256 = "1z8zhsm60hvrrbj56qfbqd1qrj2db7222r2jifz06jhzfj3cdpvd";
  };
  pkgs-with-magui = import nixpkgs-with-magui { system = "x86_64-linux"; };
in
{
  disabledModules = [ "programs/mouse-actions.nix" ];
  imports = [ "${nixpkgs-with-magui}/nixos/modules/programs/mouse-actions.nix" ];

  programs.mouse-actions = {
    enable = true;
    autorun = true;
    package = pkgs-with-magui.mouse-actions-gui;
  };
}
```

Place this file next to your configs, name it something like `magui.nix` and the import it from your configuration with
```nix
{
  imports = [ ./magui.nix ];
}
```

**Comment by waldman** [comment]
Mate... It WORKS! Perfectly! :heart:

**Review by Artturin — APPROVED** [review]
Resolved conflict

**Pull request merged**